### PR TITLE
planner, executor: check col.OrigColName instead of col.ColName in isCoveringIndex

### LINKED ISSUE #9636: we should use col.OriginColName instead col.ColName when extracting index filter
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
``` sql
drop table if exists t;
CREATE TABLE `t` (a int, KEY (a));

SELECT * FROM (SELECT * FROM 
    (SELECT a as d FROM t WHERE a IN ('100')) AS x 
WHERE x.d < "123" ) tmp_count;
```

2. What did you expect to see?
no error happens

3. What did you see instead?
panic happens
``` go
[2019/03/11 14:49:00.840 +08:00] [ERROR] [conn.go:537] ["connection running loop panic"] [conn=2] [lastCmd="SELECT * FROM (SELECT * FROM \n    (SELECT a as d FROM t WHERE a IN ('100')) AS x \nWHERE x.d < \"123\" ) tmp_count"] [err="\"invalid memory address or nil pointer dereference\""] [stack="goroutine 202 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x53f5ea0, 0xc00731a6f0, 0xc000311e10)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:535 +0x10d\npanic(0x5011760, 0x627e280)\n\t/Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/panic.go:513 +0x1b9\ngithub.com/pingcap/tidb/planner/core.(*copTask).count(0xc0088499b0, 0x8)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/task.go:65 +0x35\ngithub.com/pingcap/tidb/planner/core.(*PhysicalIndexScan).addPushedDownSelection(0xc008862000, 0xc0088499b0, 0xc007a59e00, 0x7fefffffffffffff, 0xc008070840)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:583 +0xbc\ngithub.com/pingcap/tidb/planner/core.(*DataSource).convertToIndexScan(0xc007a59e00, 0xc0080cf720, 0xc00884ae60, 0x1, 0x4, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:524 +0x589\ngithub.com/pingcap/tidb/planner/core.(*DataSource).findBestTask(0xc007a59e00, 0xc0080cf720, 0x5400a60, 0x627f910, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:404 +0x421\ngithub.com/pingcap/tidb/planner/core.(*baseLogicalPlan).findBestTask(0xc008858128, 0xc0080cf6d0, 0x0, 0x0, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:118 +0x460\ngithub.com/pingcap/tidb/planner/core.physicalOptimize(0x541f060, 0xc008858120, 0xc008858120, 0x541f060, 0xc008858120, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:150 +0xa0\ngithub.com/pingcap/tidb/planner/core.DoOptimize(0x51, 0x541f060, 0xc008858120, 0x541f060, 0xc008858120, 0x1, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:107 +0x7e\ngithub.com/pingcap/tidb/planner.Optimize(0x541dfe0, 0xc00742ab60, 0x53f76e0, 0xc008070480, 0x541bea0, 0xc008803470, 0x0, 0x0, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/optimize.go:68 +0x2ec\ngithub.com/pingcap/tidb/executor.(*Compiler).Compile(0xc007e9f508, 0x53f5de0, 0xc0080b5d80, 0x53f9660, 0xc008070480, 0x0, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/compiler.go:49 +0x1f3\ngithub.com/pingcap/tidb/session.(*session).execute(0xc00742ab60, 0x53f5de0, 0xc0080b5d80, 0xc000159ce1, 0x6f, 0x1, 0x54087c0, 0x62f6c08, 0x30, 0x8)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:948 +0x62c\ngithub.com/pingcap/tidb/session.(*session).Execute(0xc00742ab60, 0x53f5de0, 0xc0080b5d80, 0xc000159ce1, 0x6f, 0x0, 0x0, 0x0, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:910 +0xc4\ngithub.com/pingcap/tidb/server.(*TiDBContext).Execute(0xc007e7c630, 0x53f5de0, 0xc0080b5d80, 0xc000159ce1, 0x6f, 0xc007e9f788, 0x4e31131, 0xc0001a9ce0, 0x0, 0x62d6bc0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/driver_tidb.go:246 +0x7c\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc000311e10, 0x53f5de0, 0xc0080b5d80, 0xc000159ce1, 0x6f, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:1024 +0x8e\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc000311e10, 0x53f5ea0, 0xc008848d50, 0xc000159ce1, 0x70, 0x70, 0x0, 0x0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:752 +0x59c\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc000311e10, 0x53f5ea0, 0xc00731a6f0)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:588 +0x2d1\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc000177d50, 0x5403340, 0xc0000c7668)\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/server.go:408 +0x34d\ncreated by github.com/pingcap/tidb/server.(*Server).Run\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/server.go:333 +0x53d\n"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:537\nruntime.call32\n\t/Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/asm_amd64.s:522\nruntime.gopanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/panic.go:513\nruntime.panicmem\n\t/Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/panic.go:82\nruntime.sigpanic\n\t/Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/signal_unix.go:390\ngithub.com/pingcap/tidb/planner/core.(*copTask).count\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/task.go:65\ngithub.com/pingcap/tidb/planner/core.(*PhysicalIndexScan).addPushedDownSelection\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:583\ngithub.com/pingcap/tidb/planner/core.(*DataSource).convertToIndexScan\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:524\ngithub.com/pingcap/tidb/planner/core.(*DataSource).findBestTask\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:404\ngithub.com/pingcap/tidb/planner/core.(*baseLogicalPlan).findBestTask\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:118\ngithub.com/pingcap/tidb/planner/core.physicalOptimize\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:150\ngithub.com/pingcap/tidb/planner/core.DoOptimize\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:107\ngithub.com/pingcap/tidb/planner.Optimize\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/optimize.go:68\ngithub.com/pingcap/tidb/executor.(*Compiler).Compile\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/compiler.go:49\ngithub.com/pingcap/tidb/session.(*session).execute\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:948\ngithub.com/pingcap/tidb/session.(*session).Execute\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:910\ngithub.com/pingcap/tidb/server.(*TiDBContext).Execute\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/driver_tidb.go:246\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:1024\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:752\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:588\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/server.go:408"]

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB
`0bec639`

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
fix #9636 

### What is changed and how it works?
Checking col.OrigColName instead of col.ColName in isCoveringIndex

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -423,7 +423,8 @@ func isCoveringIndex(columns []*expression.Column, indexColumns []*model.IndexCo
 		isIndexColumn := false
 		for _, indexCol := range indexColumns {
 			isFullLen := indexCol.Length == types.UnspecifiedLength || indexCol.Length == col.RetType.Flen
-			if col.ColName.L == indexCol.Name.L && isFullLen {
+			// We use col.OrigColName instead of col.ColName
```
Comment (winoros) [inline planner/core/find_best_task.go]: I think adding the issue link here can help others to understand.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -423,7 +423,8 @@ func isCoveringIndex(columns []*expression.Column, indexColumns []*model.IndexCo
...
+			if col.OrigColName.L == indexCol.Name.L && isFullLen {
```
Comment (zz-jason) [inline planner/core/find_best_task.go]: Is there other places that use `col.ColName.L` instead of `col.OrigColName.L`?

Comment (XuHuaiyu) [reply planner/core/find_best_task.go]: I'm not sure.
It needs to be checked.

File: planner/core/find_best_task.go (hunk 1 above)
Comment (XuHuaiyu) [reply planner/core/find_best_task.go]: ok

File: planner/core/find_best_task.go (hunk 2 above)
Comment (eurekaka) [reply planner/core/find_best_task.go]: Can we use `UniqueID` here for this case?

Comment (winoros) [reply planner/core/find_best_task.go]: No...  There's no place store the result converting `model.IndexInfo.Columns` to `expression.Columns`. So in many places we use `model.IndexColumn` directly compares with `expression.Column`

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @winoros

**Pull request merged**